src/TypeSpec/__Private/UntypedArraySpec.hack

```diff
@@ -24,7 +24,7 @@ final class UntypedArraySpec extends TypeSpec<array> {
         TypeCoercionException::withValue($this->getTrace(), 'array', $value);
     }
 
-    $out = [];
+    $out = darray[];
```

Comment:
Technically this changes behavior, maybe we should FIXME it instead? like everything else in this file :(

I wonder if we could just delete this whole file, it's only going to cause more and more trouble as HHVM internals change from `array` to varray/darray/varray_or_darray...

@fredemott what do you think?

Reply:
`array` now /is/ `varray_or_darray`; perhaps we should switch this to literally trying both the varrayspec and darrayspec and returning success in either case?

Reply:
OK so we probably also want to change `DictLikeArraySpec` to use `darray` and `VecLikeArraySpec` to use `varray` (they're both still using `array`). I can do a separate PR for that.

Reply:
Maybe rename them too now that they're a real type? :)

Reply:
`varray` keys must be sequential, so I assumed that, since we can't guarantee that here, we couldn't do any better.

Btw, isn't that `as arraykey` going the throw when you pass in something KeyedTraversable<non_arraykey, mixed>`?

Reply:
So, things are more complicated than I realized. As far as I can tell the current unification efforts only affect the typechecker.

At runtime, `array`, `varray` and `darray` are still 3 different types. The functions `is_varray` and `is_darray` return false for an `array`, while `is_array` returns true for everything. There's also still `TypeStructure::OF_ARRAY`, `OF_VARRAY` and `OF_DARRAY` and `OF_VARRAY_OR_DARRAY`.

The typechecker considers `array<T>` the same as `varray<T>`, `array<Tk, Tv>` the same as `darray<Tk, Tv>` and `array` the same as `varray_or_darray<_>`. Also as far as I can tell, empty `array()` is considered a different, separate type.

I'm not sure what the right thing to do here is -- to be able to handle all values, we'd probably need to support both `VArrayTypeSpec` and `VecLikeArrayTypeSpec` (and similarly for DArray/DictLikeArray) but I'm not sure if that's worth adding, maybe it's better to wait until things are more stable and consistent...

Reply:
I thought `KeyedTraversable<non_arraykey, _>` wasn't possible but now I'm not so sure :/

Perhaps using `ArrayKeySpec` to coerce the key would be the right thing to do here.